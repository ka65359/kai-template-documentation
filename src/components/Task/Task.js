/**
 * @module Task
 * @exports Task
 * @description A module that builds a Task widget.
 * @author Kai
 * @version 1.0.0
 */
import React from "react";
import PropTypes from "prop-types";
import { text } from "@storybook/addon-knobs";

/**
 * @typedef {Object} Task
 *
 * @property {Object} task - Task specific details
 * @property {string} task.id - Task ID
 * @property {string} task.title - Task description
 * @property {string} task.state - Task state (checked or unchecked)
 * @property {function} onArchiveTask - Callback of what happens when a task is archived
 * @property {function} onPinTask - Callback of what happens when a task is pinned
 */
const Task = ({ task: { id, title, state }, onArchiveTask, onPinTask }) => {
  /**
   * Build the Task HTML (if it isn't archived).
   *
   * @method buildCurrentTasks
   */
  const buildCurrentTasks = () => {
    if (state !== "TASK_ARCHIVED") {
      return (
        <a href onClick={() => onPinTask(id)}>
          <span className={`icon-star`} />
        </a>
      );
    }
  };

  return (
    <div className={`list-item ${state}`}>
      <label className="checkbox">
        <input
          type="checkbox"
          defaultChecked={state === "TASK_ARCHIVED"}
          disabled={true}
          name="checked"
        />
        <span className="checkbox-custom" onClick={() => onArchiveTask(id)} />
      </label>
      <div className="title">
        <input
          type="text"
          value={text("title", title)}
          readOnly={true}
          placeholder={"Input title"}
        />
      </div>

      <div className="actions" onClick={(event) => event.stopPropagation()}>
        {buildCurrentTasks()}
      </div>
    </div>
  );
};

Task.propTypes = {
  task: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    state: PropTypes.string.isRequired
  }),
  onArchiveTask: PropTypes.func,
  onPinTask: PropTypes.func
};

export default Task;
