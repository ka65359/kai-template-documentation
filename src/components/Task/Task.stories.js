import React from "react";
import { action } from "@storybook/addon-actions";
import { withKnobs } from "@storybook/addon-knobs";

import Task from "./Task";

export default {
  component: Task,
  title: "Task",
  decorators: [withKnobs],
  // Our exports that end in "Data" are not stories.
  excludeStories: /.*Data$/
};

export const taskData = {
  id: "1",
  title: "Test Task",
  state: "TASK_INBOX",
  updatedAt: new Date(2018, 0, 1, 9, 0)
};

export const actionsData = {
  onPinTask: action("onPinTask"),
  onArchiveTask: action("onArchiveTask")
};

export const Default = () => <Task task={{ ...taskData }} {...actionsData} />;

export const Pinned = () => (
  <Task task={{ ...taskData, state: "TASK_PINNED" }} {...actionsData} />
);

export const Archived = () => (
  <Task task={{ ...taskData, state: "TASK_ARCHIVED" }} {...actionsData} />
);
