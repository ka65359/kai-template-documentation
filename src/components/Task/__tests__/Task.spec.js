import React from "react";
import { shallow } from "enzyme";
import Task from "../Task";

describe("Task() tests", () => {
  let onArchiveTask;
  let onPinTask;
  let data;

  beforeEach(function() {
    onArchiveTask = jest.fn();
    onPinTask = jest.fn();
    data = {
      task: {
        id: 0,
        title: "Sample Title",
        state: "TASK_PINNED"
      },
      onArchiveTask,
      onPinTask
    };
  });

  afterEach(function() {
    onPinTask.mockReset();
    onPinTask = null;
    onArchiveTask.mockReset();
    onArchiveTask = null;
    data = null;
  });

  it("should render correctly in mode", () => {
    const component = shallow(<Task {...data} />);

    expect(component).toMatchSnapshot();
  });
});
