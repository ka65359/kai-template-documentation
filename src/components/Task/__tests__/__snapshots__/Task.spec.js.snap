// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task() tests should render correctly in mode 1`] = `ShallowWrapper {}`;
